rust/src/client.rs

```diff
@@ -0,0 +1,197 @@
+    /// `id`: an ID to assign to the `Client`.
+    /// Returns `Ok(client)` if `Client` `client` initialised successfully
+    /// Returns `Err(err)` if `ClientError` `err` occurred
+    pub fn new_with_id(period: u64, handle: Handle, id: u32) -> Result<Self, ClientError> {
+        let participant = Participant::new()
+            .map_err(ClientError::ParticipantInitErr)?;
+        Ok(Self {
+            handle,
+            participant,
+            interval: time::interval(Duration::from_secs(period)),
+            id,
+        })
+    }
+
+    /// Start the `Client` loop
+    ///
+    /// Returns `Err(err)` if `ClientError` `err` occurred
+    /// NOTE in the future this may iterate only a fixed number of times, at the
+    /// end of which this returns `Ok(())`
+    pub async fn start(&mut self) -> Result<(), ClientError> {
+        loop {
+            // any error that bubbles up will finish off the client
+            self.per_round().await?;
+        }
+    }
+
+    /// Client duties within a round
+    pub async fn per_round(&mut self) -> Result<Task, ClientError> {
+        let round_params: Arc<RoundParameters> = loop {
+            if let Some(round_params) =
+                self.handle.get_round_parameters().await
+            {
+                break round_params
+            }
+            println!("{}: round params not ready yet, retrying in a sec...", self.id);
+            self.interval.tick().await;
+        };
+        let round_seed: &[u8] = round_params.seed.as_slice();
+        println!("computing sigs and checking task");
+        self.participant
+            .compute_signatures(round_seed);
+        let (sum_frac, upd_frac) = (round_params.sum, round_params.update);
+        match self
+            .participant
+            .check_task(sum_frac, upd_frac)
+        {
+            Task::Sum    =>
+                self.summer(round_params.pk).await,
+            Task::Update =>
+                self.updater(round_params.pk).await,
+            Task::None   =>
+                self.unselected().await,
+        }
+    }
+
+    /// Duties for unselected clients
+    async fn unselected(&mut self) -> Result<Task, ClientError> {
+        Ok(Task::None)
+    }
+
+    /// Duties for clients selected as summers
+    async fn summer(&mut self, coord_pk: CoordinatorPublicKey)
+                    -> Result<Task, ClientError>
+    {
+        let sum1_msg: Vec<u8> = self
+            .participant
+            .compose_sum_message(&coord_pk);
+        self.handle
+            .send_message(sum1_msg)
+            .await;
+
+        let pk = self.participant.get_encr_pk();
+        let seed_dict_ser: Arc<Vec<u8>> = loop {
+            if let Some(seed_dict_ser) = self.handle.get_seed_dict(pk).await {
+                break seed_dict_ser
+            }
+            // updates not yet ready, try again later...
+            self.interval.tick().await;
+        };
```

Comment:
it seems to me that this might lead to stale `seed_dict`s and therefore ignored messages. if we consider the following:
- in round `i` the participant successfully awaits the `round_params` in `per_round()` and is assigned the `Sum` task
- it then attempts to await the `summer()` and in the meantime the coordinator decides to restart the round before the `seed_dict` is broadcasted
- in round `i+1` the two previous steps are repeated, but now a `seed_dict` is broadcasted and successfully awaited
- then both loops from round `i` and `i+1` terminate and two `sum2` messages are composed and sent, but the message referencing the `round_params` from `i` will of course be ignored by the coordinator in round `i+1` resulting in unnecessary overhead for both participant and coordinator

i think we need to keep track if a round has changed (via the coordinator pk in the round params). for example we can introduced a round counter that gets incremented on a new coordinator pk (this round counter doesn't reflect the real round number of the federated learning, but that's not needed anyways). then in the loop or the futures `poll()` (not sure how to do this best) we can check for that abort the loop and the whole method if needed.

one main question would be if this scenario is common or if it's rare and we can accept the occasional overhead? the larger the time frames for the federated learning, the rarer it probably is.

Reply:
that's a fair point; at the moment we don't consider the possibility of the coordinator aborting mid-round. given that that effects are probably not devastating (maybe just unnecessary overhead), I'd say we can postpone a solution to this (if indeed, one is really required). I'll add an issue for it.

Reply:
i agree that this is a topic for future optimization.